simapp/abci.go

```diff
@@ -2,69 +2,179 @@ package simapp
 
 import (
 	"bytes"
-	"crypto/rand"
 	"encoding/json"
 	"fmt"
 
 	abci "github.com/cometbft/cometbft/abci/types"
 
 	"github.com/cosmos/cosmos-sdk/baseapp"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
 )
 
+/*
+1. PrepareProposal -> Verify Prev. Block's VEs (check for 2/3+ majority ) -> Encode VEs in tx bytes
+2. ProcessProposal -> WE CAN execute txs -> resp. tx handler will set data buffer (on which we will eventually vote upon). (ALTERNATIVELY maybe we can do this in FinalizeBlock ?)
+3. ExtendVote -> Fetch data from buffer -> Vote on it acc. -> Store result.
+4. VerifyVoteExtension -> Verify Vote recieved from other validators and ACCEPT/REJECT acc.
+5. PreFinalizeBlock -> Extract VEs from prev. block and persist in DB.
+6. FinalizeBlock -> Finalize and execute txs.
+7. CommitBlock -> Self explanatory.
+*/
+
 type (
 	// VoteExtensionHandler defines a dummy vote extension handler for SimApp.
 	//
 	// NOTE: This implementation is solely used for testing purposes. DO NOT use
 	// in a production application!
-	VoteExtensionHandler struct{}
+	VoteExtensionHandler struct {
+		app *SimApp
+	}
 
 	// VoteExtension defines the structure used to create a dummy vote extension.
 	VoteExtension struct {
 		Hash   []byte
 		Height int64
 		Data   []byte
+		Result int
 	}
 )
 
-func NewVoteExtensionHandler() *VoteExtensionHandler {
-	return &VoteExtensionHandler{}
+const (
+	Vote_YES = iota
+	Vote_NO
+	Vote_SKIP
+)
+
+func NewVoteExtensionHandler(app *SimApp) *VoteExtensionHandler {
+	return &VoteExtensionHandler{app: app}
 }
 
-func (h *VoteExtensionHandler) SetHandlers(bApp *baseapp.BaseApp) {
-	bApp.SetExtendVoteHandler(h.ExtendVote())
-	bApp.SetVerifyVoteExtensionHandler(h.VerifyVoteExtension())
+// PrepareProposal -> Verify Prev. Block's VEs (check for 2/3+ majority ) -> Encode VEs in tx bytes
+func (app *SimApp) NewPreparePropsalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+
+		event, err := app.StakingKeeper.GetEventRecord(ctx)
+		if err != nil {
+			fmt.Println("Error occured while fetching Event Record from DB: ", err)
+		}
+		fmt.Println("Event stored in height : ", ctx.BlockHeight()-1, "event ID: ", event.Id, "event block no.: ", event.BlockNumber)
+
+		// Validate VE sigs and check whether they have 2/3+ majority
+		hasTwoThirdsSigs := true
+		if err := app.StakingKeeper.ValidateVoteExtensions(ctx, ctx.BlockHeight(), ctx.ChainID(), req.LocalLastCommit); err != nil {
+			fmt.Println("PrepareProposal: Error occured while validating VEs: ", err)
+			hasTwoThirdsSigs = false
+		}
+
+		txs := req.Txs
+		// Only include VE if there's majority
+		if hasTwoThirdsSigs {
+			fmt.Println("PrepareProposal: INJECTING VEs!")
+
+			for _, v := range req.LocalLastCommit.Votes {
```

Comment:
A malicious node could censor side transactions here and nodes executing processProposal would not have a means to detect it.

Reply:
Noted, we'll modify the approach in the updated PoC.

Reply:
Hey @lasarojc , if the malicious node were to send empty or wrong VEs, would there be any way that the other nodes can detect it ? I guess they can query their local state and catch but is it possible to do it on the app level or will the Comet handle it ?